physx-sys/src/physx_api.cpp

```diff
@@ -52,45 +52,73 @@ PxFilterFlags FilterShaderTrampoline(PxFilterObjectAttributes attributes0,
     return PxFilterFlags{shaderfilter(&info)};
 }
 
-class CollisionFilterTrampoline : public PxSimulationEventCallback
+typedef void (*CollisionCallback)(void *, PxContactPairHeader const *, PxContactPair const *, PxU32);
+typedef void (*TriggerCallback)(void *, PxTriggerPair const *, PxU32);
+typedef void (*ConstraintBreakCallback)(void *, PxConstraintInfo const *, PxU32);
+typedef void (*WakeSleepCallback)(void *, PxActor **const, PxU32, bool);
+typedef void (*AdvanceCallback)(void *, const PxRigidBody *const *, const PxTransform *const, PxU32);
+
+struct SimulationEventCallbackInfo {
+    CollisionCallback collisionCallback;
+    void *collisionUserData;
+    TriggerCallback triggerCallback = nullptr;
+    void *triggerUserData = nullptr;
+    ConstraintBreakCallback constraintBreakCallback = nullptr;
+    void *constraintBreakUserData = nullptr;
+    WakeSleepCallback wakeSleepCallback = nullptr;
+    void *wakeSleepUserData = nullptr;
+    AdvanceCallback advanceCallback = nullptr;
+    void *advanceUserData = nullptr;
+};
+
+class SimulationEventTrampoline : public PxSimulationEventCallback
 {
   public:
-    CollisionFilterTrampoline(CollisionCallback callback, void *userData) : mUserData(userData), mCallback(callback)
-    {}
-
-    void onContact(const PxContactPairHeader &pairHeader, const PxContactPair *pairs, PxU32 nbPairs) override
-    {
-        mCallback(mUserData, &pairHeader, pairs, nbPairs);
+    SimulationEventTrampoline(const SimulationEventCallbackInfo *callbacks) {
+        mCallbacks = *callbacks;
```

Comment:
Do this in the initializer list instead, not in the body.

Reply:
To be fair it really irks me overall that this is just copying the whole struct - I get why; but this is essentially/potentially a transfer of ownership to this callback... I'm not sure if we can use lifetimes here but really the SimulationEventCallbackInfo has to have a lifetime at least as long as the trampoline itself... Which are both less than the full scene. Hmm. I think we've had this discussion before; and I still don't know how to represent it in Rust terms.

Reply:
It's a small struct and it's easier to just copy it. This is C++ so no lifetime annotations...

Reply:
I know it's C++, my point was more that I'd want this to be safer and be more transparent w.r.t. the expected lifetimes on the Rust side - making the API safer. :-) Maybe something funky like

```rust
enum CallbackPayload {
    OnWake(...),
    OnContact(...),
}

struct SimulationCallback<'a, WakeUserData, ContactUserData, ...> {
     callback_trampoline: *mut PxSimulationCallback,
     wake_data: SomeRefType<'a, WakeUserData>,
     ...
}

impl SimulationCallback<...> {
    fn on_callback(self: *mut Self, data: *const CallbackPayload) { ... }
}
```

And then ~ish the callback trampoline is tied to some lifetime in Rust land... Just thinking out loud.

Reply:
Figuring out how to do this stuff in a safe or safer manner is not something I have time or energy for right now though, unfortunately. There is heavy demand for trigger volumes and a bunch more features and this works for now.

After all this is in place, I want to go back and move all the unsafe stuff we've been doing in Ark into safe nice wrappers here, but it hasn't been possible to do it continuously since the speed of change and discovery of more stuff we need from PhysX has just been too high. We need actual usage to know what the safe interfaces should look like anyway.

Reply:
Yeah, I understand that! Approved now.